<a href="https://colab.research.google.com/github/KacperKaszuba0608/Projects/blob/main/Exploring_eBay_Car_Sales_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Guided Project: Exploring eBay Car Sales Data

In this project we focused on exploring eBay Car Sales Data from Kaggle where the original dataset isn't availble now, but you can find it on [data.word](https://data.world/data-society/used-cars-data). Dataquest team sampled 50000 rows and change some values. They did it because we tried clean missing or understood values and then explore them. Below you have description of columns:

* `dateCrawled` - when this ad was first crawled. All field-values are taken from this date
* `name` - name of the car
* `seller` - private or dealer
* `offerType` - the tipe of listing
* `price` - the price on the ad to sell the car
* `abtest` - whether the listing is included in an A/B test
* `vehicleType`
* `yearOfRegistration` - at which year the car was first registered
* `gearbox` - the transmisson type
* `powerPS` - power of the car in PS
* `model` - the car model name
* `kilometer` - how many kilometers the car has driven
* `monthOfRegistration` - at which month the car was first registered
* `fuelType`
* `brand`
* `notRepairedDamage` - if the car has a damage which is not repaired yet
* `dateCreated` - the date for which the ad at ebay was created
* `nrOfPictures` - number of pictures in the ad
* `postalCode` - the postal code for the location of the vehicle
* `lastSeen` - when the crawler saw this ad last online

## Importing libraries and dataset

In [75]:
import pandas as pd
import numpy as np

cars = pd.read_csv('https://query.data.world/s/b7vnijn6k5r4awstrt4xkouqwff3mi', encoding = 'Latin-1')

In [76]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [77]:
cars.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

As we see above the dataset has 50000 rows and 20 columns. Most of them are `object` type and only 5 columns have `int64` type. 5 columns have null values, so we had to clean it and some of column names are no in the same type. That's we changed them below.

## Cleaning Column Names

In [78]:
cars.columns = ['crawled_date', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_num', 'postal_code',
       'last_seen']
cars.head()

crawled_date                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   odometer  registration_month fuel_type       brand unrepaired_damage  \
0    150000                   0    benzin  volkswagen               NaN   
1    125000                   5    diesel        audi                ja   
2    125000                   8    diesel        jeep               NaN   
3    150000                   6    benzin  volkswagen              nein   
4     90000                   7    diesel       skoda              nein   

            ad_created  pictures_num  postal_code            last_seen  
0  2016-03-24 00:00:00             0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0        60437  2016-04-06 10:17:21

Now, column names are more simple and understood. We created list of new column names and asigned it to `cars.columns` function from pandas.

## Exploration and Cleaning

In [79]:
cars.describe(include = 'all')

crawled_date         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model       odometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  pictures_num   postal_code  \
count             299468               371528      371528.0  371528.00000   
unique                 2                  114           NaN           NaN   
top                 nein  2016-04-03 00:00:00           NaN           NaN   
freq              263182                14450           NaN           NaN   
mean                 NaN                  NaN           0.0   50820.66764   
std                  NaN                  NaN           0.0   25799.08247   
min                  NaN                  NaN           0.0    1067.00000   
25%                  NaN                  NaN           0.0   30459.00000   
50%                  NaN                  NaN           0.0   49610.00000   
75%                  NaN                  NaN           0.0   71546.00000   
max                  NaN                  NaN           0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Description of columns tels us that 2 columns: `odometer` and `price` have combination of string and number. That's why we had to found a pattern and cleaned it. Column `registration_year` has a strange maximum value equal 9999. It's probably wrong value because 9999 year is a future. Column `power_ps` has unbelievable maximum value qual 177700 ps. None car has power like this one. Column `registration_month` has minimum value 0.00, month 0 doesn't exist so it's misstake too. We colud remove columns which every value were the same. It happened in columns: `seller`, `offer_type` and `pictures_num`. They told us nothing so we could remove it. What we did at first.

In [80]:
cars = cars.drop(['seller', 'offer_type', 'pictures_num'], axis = 1)
cars.shape

(371528, 17)

In [81]:
cars['price']

0           480
1         18300
2          9800
3          1500
4          3600
          ...  
371523     2200
371524     1199
371525     9200
371526     3400
371527    28990
Name: price, Length: 371528, dtype: int64

Data which we use in this project from Dataquest has signs: "$", commas or dots". So we had to replace it. In data from data.world we didn't have to do this because columns `price` and `odometer` had have already numeric type.

In [82]:
#cars['price'] = cars['price'].str.replace('$', '').str.replace(',', '')
#cars['price'] = cars['price'].astype(int)

In [83]:
#cars['odometer'] = cars['odometer'].str.replace('km', '').str.replace(',', '')
#cars['odometer'] = cars['odometer'].astype(int)

In [84]:
cars.rename({'odometer' : 'odometer_km', 'price' : 'price_dol'}, 
            inplace = True, axis = 1)

## Exploring and cleaning prices

In [85]:
cars['price_dol'].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price_dol, dtype: float64

In [86]:
cars['price_dol'].value_counts().sort_index(ascending = False).head(20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price_dol, dtype: int64

In [87]:
cars['price_dol'].value_counts().sort_index(ascending = True).head(20)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price_dol, dtype: int64

As we see above we have cars which cost 0\\$ and cars which cost 999990$ and more. That's why we chose cars' prices between 1\\$ and 350001\\$. We did it below.

In [88]:
cars = cars[cars['price_dol'].between(1, 350001)]
cars['price_dol'].describe()

count    360635.000000
mean       5898.671956
std        8866.359669
min           1.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_dol, dtype: float64

## Exploring and cleaning odometer_km

In [89]:
cars['odometer_km'].describe()

count    360635.000000
mean     125675.364288
std       39818.609118
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

As we see odometer column is rounded and we decided that nothing in it is mistake and we didn't have to change anything.

## Exploring and cleaning registration_year and month

In [90]:
cars['registration_year'].describe()

count    360635.000000
mean       2004.433133
std          81.016977
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

Above, maximum years is 9999 and it's imposible. Minimum value is 1000 and it historically imposible too. That's we had to find registration year between 1886 (date of creation first car) and now (2022).

In [91]:
cars = cars[cars['registration_year'].between(1886, 2022)]

In [92]:
cars['registration_year'].describe()

count    360487.000000
mean       2003.448124
std           7.622978
min        1910.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        2019.000000
Name: registration_year, dtype: float64

In [93]:
cars['registration_month'].describe()

count    360487.000000
mean          5.797885
std           3.683600
min           0.000000
25%           3.000000
50%           6.000000
75%           9.000000
max          12.000000
Name: registration_month, dtype: float64

In [94]:
cars.loc[cars['registration_month'] == 0, 'registration_month'].value_counts()

0    33052
Name: registration_month, dtype: int64

`registration_month` column has one mistake. It has values equal 0. All know that months numbers are 1 to 12, so we had to drop rows with 0. There were 4468 of these values.

In [95]:
cars = cars[cars['registration_month'].between(1,12)]

In [96]:
cars['registration_month'].describe()

count    327435.000000
mean          6.383136
std           3.347064
min           1.000000
25%           4.000000
50%           6.000000
75%           9.000000
max          12.000000
Name: registration_month, dtype: float64

In [97]:
(cars['registration_year'].value_counts()
                          .sort_values(ascending = False)
                          .head(15))

1999    19810
2005    19806
2006    19245
2000    19070
2004    18590
2003    18544
2001    18170
2002    17658
2007    16956
2008    15565
1998    15239
2009    15128
1997    12239
2010    12047
2011    11830
Name: registration_year, dtype: int64

The most cars registrated 20 years ago

## Exploring and cleaning power_ps

In [98]:
cars['power_ps'].describe()

count    327435.000000
mean        120.874302
std         181.855225
min           0.000000
25%          75.000000
50%         110.000000
75%         150.000000
max       20000.000000
Name: power_ps, dtype: float64

The fastest car in the world has 1600ps, so this to upper limit of ps value, that why every value over it has to be remove. The car can't have power qual 0, so we had to check it too.

In [99]:
cars['power_ps'].value_counts().sort_index().head(20)

0     24002
1        22
2         7
3         8
4        28
5        89
6        11
7        10
8         7
9         3
10       14
11       27
12       16
13        6
14       18
15       12
16        5
17        7
18       48
19       11
Name: power_ps, dtype: int64

In [100]:
cars.loc[cars['power_ps'] > 1600, 'power_ps'].value_counts().sum()

144

In [101]:
cars = cars[cars['power_ps'].between(1, 1601)]

In [102]:
cars['power_ps'].describe()

count    303289.000000
mean        127.563400
std          66.665768
min           1.000000
25%          80.000000
50%         116.000000
75%         150.000000
max        1600.000000
Name: power_ps, dtype: float64

In [103]:
cars['power_ps'].value_counts().sort_values(ascending = False).head(10)

75     21122
60     13975
150    13972
140    12957
101    11948
116    11201
90     11077
170    10289
105     9928
136     6379
Name: power_ps, dtype: int64

The most cars have power between 60 and 150 ps.

## Exploring the date columns

In [104]:
cars[['crawled_date','ad_created','last_seen']][0:5]

crawled_date           ad_created            last_seen
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21
5  2016-04-04 17:36:23  2016-04-04 00:00:00  2016-04-06 19:17:07

In [105]:
(cars['crawled_date'].str[:10]
                     .value_counts(normalize = True, dropna = False)
                     .sort_index())

2016-03-05    0.026068
2016-03-06    0.014616
2016-03-07    0.035867
2016-03-08    0.033285
2016-03-09    0.033875
2016-03-10    0.032164
2016-03-11    0.032425
2016-03-12    0.036774
2016-03-13    0.016051
2016-03-14    0.036140
2016-03-15    0.032906
2016-03-16    0.029876
2016-03-17    0.031340
2016-03-18    0.013070
2016-03-19    0.035128
2016-03-20    0.036553
2016-03-21    0.035455
2016-03-22    0.032190
2016-03-23    0.031884
2016-03-24    0.029731
2016-03-25    0.032761
2016-03-26    0.032312
2016-03-27    0.030304
2016-03-28    0.035554
2016-03-29    0.033961
2016-03-30    0.033433
2016-03-31    0.031884
2016-04-01    0.034433
2016-04-02    0.035247
2016-04-03    0.039444
2016-04-04    0.037904
2016-04-05    0.012533
2016-04-06    0.003211
2016-04-07    0.001622
Name: crawled_date, dtype: float64

In [106]:
(cars['last_seen'].str[:10]
                     .value_counts(normalize = True, dropna = False)
                     .sort_index())

2016-03-05    0.001213
2016-03-06    0.003993
2016-03-07    0.005028
2016-03-08    0.007442
2016-03-09    0.009367
2016-03-10    0.010970
2016-03-11    0.012463
2016-03-12    0.022896
2016-03-13    0.008098
2016-03-14    0.011873
2016-03-15    0.015780
2016-03-16    0.016011
2016-03-17    0.027993
2016-03-18    0.006782
2016-03-19    0.015731
2016-03-20    0.019529
2016-03-21    0.019473
2016-03-22    0.020073
2016-03-23    0.017620
2016-03-24    0.018685
2016-03-25    0.018458
2016-03-26    0.015540
2016-03-27    0.016347
2016-03-28    0.021920
2016-03-29    0.022866
2016-03-30    0.023390
2016-03-31    0.023901
2016-04-01    0.023730
2016-04-02    0.024841
2016-04-03    0.025075
2016-04-04    0.025289
2016-04-05    0.129748
2016-04-06    0.223968
2016-04-07    0.133905
Name: last_seen, dtype: float64

As we see above, columns `crawled_date` and `last_seen` are percentages similar. The ads were crawled day by day for two months (March and April). Three last rows in `last_seen` column are diferent and disturb the rest of distribution. This values are higher than earlier but it's not because of sale many car, but probably bacause ending the crawled.

In [107]:
(cars['ad_created'].str[:10]
                     .value_counts(normalize = True, dropna = False)
                     .sort_index())

2015-08-07    0.000003
2015-08-10    0.000003
2015-09-04    0.000007
2015-09-09    0.000003
2015-10-14    0.000003
                ...   
2016-04-03    0.039619
2016-04-04    0.037980
2016-04-05    0.011375
2016-04-06    0.003202
2016-04-07    0.001553
Name: ad_created, Length: 110, dtype: float64

The most ads were created in two months (March and April) and that months tally with `crawled_date` and `last_seen` columns. But in `ad_created` column we saw records which had ads created 9 month earlier than first record in `crawled_date` column.

## Exploring Price by brand

We chose top 20 brands and explore mean price of car.

In [108]:
def means_by_brand(dataset, brand_column, mean_column):
    '''
    dataset - variable with data (dataframe)
    brand_column - name of column with brands (string)
    mean_column - name of column witch we want calcuate mean (string)
    '''
    mean_by_brands = {}
    
    brands = dataset[brand_column].value_counts(normalize = True).head(20).index
    
    for brand in brands:
        one_brand = dataset[dataset[brand_column] == brand]
        mean_value = one_brand[mean_column].mean()
        mean_by_brands[brand] = int(mean_value)
        
    return mean_by_brands

In [109]:
mean_price = means_by_brand(cars, 'brand', 'price_dol')
mean_price

{'volkswagen': 5849,
 'bmw': 8922,
 'opel': 3280,
 'mercedes_benz': 9023,
 'audi': 9646,
 'ford': 4063,
 'renault': 2718,
 'peugeot': 3498,
 'fiat': 3185,
 'seat': 4847,
 'skoda': 6840,
 'mazda': 4439,
 'smart': 3777,
 'citroen': 3991,
 'toyota': 5535,
 'nissan': 5157,
 'mini': 10088,
 'hyundai': 5894,
 'volvo': 5630,
 'mitsubishi': 3704}

Top 5 most expensive cars are:
1. mini
2. audi
3. mercedes_benz
4. bmw
5. skoda

So we can conclude that:
* Audi, mercedes and bwm are expensive and luxury cars. 
* Opel and ford are chipper cars but still good.
* Skoda and volkswagen are the best option for both fiancial status. 

We deduct some brands in conclusion because some of them has low value of percentage of the total values.

## Storing Aggregate Data in a DataFrame

Below we created new data frame which store aggraagate data about mean price and mean mileage for top 20 brands.

In [110]:
mileage = means_by_brand(cars, 'brand', 'odometer_km')

In [111]:
brand_info = pd.DataFrame(pd.Series(mean_price), 
                          columns=['mean_price'])
brand_info['mileage'] = pd.Series(mileage)

In [112]:
brand_info

mean_price  mileage
volkswagen           5849   127649
bmw                  8922   132648
opel                 3280   127886
mercedes_benz        9023   130356
audi                 9646   128608
ford                 4063   123202
renault              2718   126540
peugeot              3498   123347
fiat                 3185   115396
seat                 4847   119162
skoda                6840   112733
mazda                4439   124527
smart                3777    98225
citroen              3991   119497
toyota               5535   116363
nissan               5157   117896
mini                10088    94027
hyundai              5894   102561
volvo                5630   138010
mitsubishi           3704   125522

As we see above, all cars have similar mileage what confirm our earlier conclusion. And at the end of this project we can say that opel and ford are cheap cars with same mileage like 'luxury' cars (mean bmw, audi and mercedes_benz). Namely skoda and volkswagen are for both clients with high standrads and low. We can ad that skoda has lower mileage than volkswagen but volkswagen is cheaper. So, something for something.

## Translating german words to english eqivalent

In [113]:
cars[['vehicle_type', 'gearbox', 
      'model', 'fuel_type', 'unrepaired_damage']].head(10)

vehicle_type    gearbox    model fuel_type unrepaired_damage
1         coupe    manuell      NaN    diesel                ja
2           suv  automatik    grand    diesel               NaN
3    kleinwagen    manuell     golf    benzin              nein
4    kleinwagen    manuell    fabia    diesel              nein
5     limousine    manuell      3er    benzin                ja
6        cabrio    manuell  2_reihe    benzin              nein
8           bus    manuell    c_max    benzin               NaN
10    limousine    manuell  3_reihe    benzin              nein
11        kombi    manuell   passat    diesel                ja
12        kombi    manuell   passat    benzin               NaN

As we see, columns above shows strings in german. Our chalange were translating this words to english eqivalent. We did it below.

### Vehicle_type

In this column we had to change only two german words:
* kleinwagen to small cars
* andere to other

In [114]:
cars['vehicle_type'].unique()

array(['coupe', 'suv', 'kleinwagen', 'limousine', 'cabrio', 'bus',
       'kombi', nan, 'andere'], dtype=object)

In [115]:
cars.loc[cars['vehicle_type'] == 'kleinwagen', 'vehicle_type'] = 'small car'
cars.loc[cars['vehicle_type'] == 'andere', 'vehicle_type'] = 'other'

In [116]:
cars['vehicle_type'].unique()

array(['coupe', 'suv', 'small car', 'limousine', 'cabrio', 'bus', 'kombi',
       nan, 'other'], dtype=object)

### Gearbox

Here, we had to change all words to english:
* manuell to manual
* automatik to automatic

In [117]:
cars['gearbox'].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [118]:
cars.loc[cars['gearbox'] == 'manuell', 'gearbox'] = 'manual'
cars.loc[cars['gearbox'] == 'automatik', 'gearbox'] = 'automatic'

In [119]:
cars['gearbox'].unique()

array(['manual', 'automatic', nan], dtype=object)

### Car model

In this column german words are "andere" which we had to change to "other". Of course dataset has more german words but changing them will be difficult and very monotonous.

In [120]:
cars.loc[cars['model'] == 'andere', 'model'] = 'other'

### Fuel type

Here, we had the same situation like above and we just change 'andere' word.

In [121]:
cars['fuel_type'].unique()

array(['diesel', 'benzin', nan, 'lpg', 'andere', 'hybrid', 'cng',
       'elektro'], dtype=object)

In [122]:
cars.loc[cars['fuel_type'] == 'andere', 'fuel_type'] = 'other'

In [123]:
cars['fuel_type'].unique()

array(['diesel', 'benzin', nan, 'lpg', 'other', 'hybrid', 'cng',
       'elektro'], dtype=object)

### Unrepaired damage

This column has two german words ("nein", "ja"). We changed it to english "no" and "yes".

In [124]:
cars['unrepaired_damage'].unique()

array(['ja', nan, 'nein'], dtype=object)

In [125]:
cars.loc[cars['unrepaired_damage'] == 'ja', 'unrepaired_damage'] = 'yes'
cars.loc[cars['unrepaired_damage'] == 'nein', 'unrepaired_damage'] = 'no'

In [126]:
cars['unrepaired_damage'].unique()

array(['yes', nan, 'no'], dtype=object)

## How much cheaper are cars with damage than their non-damaged counterparts?

In [127]:
damaged_cars_price = round(cars.loc[cars['unrepaired_damage'] == 'yes',
                              'price_dol'].mean(), 4)
non_damaged_cars_price = round(cars.loc[cars['unrepaired_damage'] == 'no',
                              'price_dol'].mean(), 4)

print(f'Damged cars cost {damaged_cars_price}$\nNon-damaged cars cost {non_damaged_cars_price}$')

Damged cars cost 2476.2696$
Non-damaged cars cost 7342.3203$


Based on average of prices, concluding is non-damaged cars are expensive 3 times more